Release test for AWS IAM Postgres passwordless

raviharshicorp opened the pull request:
## Background

This release test tests the passwordless authentication to AWS postgres using IAM based token.

Relates OR Closes #(https://hashicorp.atlassian.net/browse/IND-4009)

JIRA: https://hashicorp.atlassian.net/browse/IND-4009

Related PRs
terraform-enterprise: https://github.com/hashicorp/terraform-enterprise/pull/3397
terraform-aws-terraform-enterprise:  https://github.com/hashicorp/terraform-aws-terraform-enterprise
ptfedev-infra: https://github.com/hashicorp/ptfedev-infra/pull/887


## How has this been tested?
The release test ran on Github:https://github.com/hashicorp/terraform-enterprise/actions/runs/19146818577/job/54726521821
Checked the TFE logs, database connection was success.



### Did you add a new setting?
No

CI/CD: https://github.com/hashicorp/terraform-enterprise/actions/runs/19464088212/job/55694964359

## Screenshot:
<img width="1511" height="824" alt="Screenshot 2025-11-20 at 19 49 04" src="https://github.com/user-attachments/assets/1ecd7ce5-05c4-4b12-addd-471175458ce8" />
<img width="1085" height="265" alt="Screenshot 2025-11-20 at 19 50 23" src="https://github.com/user-attachments/assets/540cd201-86eb-4d5b-9aad-9ff76bdc1bfa" />
<img width="1512" height="597" alt="Screenshot 2025-11-20 at 19 51 17" src="https://github.com/user-attachments/assets/ba3e294b-13f3-4c7d-aa3a-49aa5584f60e" />
<img width="1511" height="575" alt="Screenshot 2025-11-20 at 19 54 56" src="https://github.com/user-attachments/assets/2698ca0d-b4f4-4a3d-827e-795821585e53" />
<img width="1512" height="283" alt="Screenshot 2025-11-20 at 19 55 24" src="https://github.com/user-attachments/assets/7fe1a258-d41d-4c08-96ab-9ab12b07fa67" />
<img width="1512" height="629" alt="Screenshot 2025-11-20 at 19 55 47" src="https://github.com/user-attachments/assets/a07ecd66-3ba6-413a-8796-4f6800c5cbaa" />

raviharshicorp committed:
feat: add PostgreSQL AWS IAM authentication support

- Add database_passwordless_aws_use_iam and database_passwordless_aws_region variables
- Add DATABASE_AUTH_USE_AWS_IAM and DATABASE_AUTH_AWS_DB_REGION environment variables
- Update both runtime_container_engine_config and tfe_init modules
- Follow Redis passwordless authentication pattern for PostgreSQL

raviharshicorp committed:
Fix PostgreSQL IAM authentication template variable mapping

- Add Database.Passwordless.AWSUseInstanceProfile structure to tfe_init template variables
- Add database_aws_iam_auth_enabled and database_aws_iam_region variables to settings module
- This provides the correct nested structure that TFE templates expect for IAM authentication
- Templates look for .Database.Passwordless.AWSUseInstanceProfile to set DATABASE_AUTH_USE_AWS_IAM

Fixes: PostgreSQL passwordless authentication was failing because DATABASE_AUTH_USE_AWS_IAM was not being set correctly

raviharshicorp committed:
Add automated PostgreSQL IAM user creation

- Added database_iam_username variable to tfe_init module
- Updated template to use proper database connection variables
- Template now automatically creates PostgreSQL IAM user with proper permissions
- Fixed variable reference from database_user to database_iam_username
- Added comprehensive error handling and database readiness checks

This completes the automation requirements for PostgreSQL passwordless authentication release tests.

raviharshicorp committed:
Fix AWS region configuration for PostgreSQL IAM user creation

Set AWS_DEFAULT_REGION environment variable before AWS CLI commands:
- Export AWS_DEFAULT_REGION=${database_aws_iam_region}
- Ensures AWS CLI commands in PostgreSQL IAM user creation have proper region
- Resolves 'You must specify a region' AWS CLI errors

This should fix the user_data script failures and allow proper PostgreSQL IAM user creation.

raviharshicorp committed:
Add DATABASE_URL environment variable for PostgreSQL IAM authentication

TFE's auth.go code requires DATABASE_URL environment variable for IAM authentication to work:
- Added DATABASE_URL construction using database connection parameters
- Format: postgresql://user:password@host:5432/dbname?parameters
- Critical for TFE to properly use AWS IAM database authentication

This should resolve the 'database password must be set' configuration error.

raviharshicorp committed:
Fix DATABASE_URL construction for null password in IAM auth

Handle null password properly in DATABASE_URL construction:
- Only include password part if database_password is not null
- Format: postgresql://user@host:5432/dbname (no password part for IAM auth)
- Ensures valid PostgreSQL connection string for IAM authentication

This should fix DATABASE_URL format for pgmultiauth IAM authentication.

raviharshicorp committed:
Fix DATABASE_URL for PostgreSQL IAM auth: add placeholder password

TFE requires a password in the DATABASE_URL even when IAM authentication
is enabled. Added 'aws-iam-auth' placeholder password to ensure proper
DATABASE_URL format: postgresql://user:aws-iam-auth@host:5432/db?params

raviharshicorp committed:
Implement AWS RDS IAM authentication with runtime token generation

- Set DATABASE_URL to null for IAM auth (generated at runtime)
- Add RDS IAM token generation function in tfe.sh.tpl
- Generate authentication tokens using 'aws rds generate-db-auth-token'
- Inject DATABASE_URL with actual IAM token into Docker Compose at runtime
- Use IAM username and token-based authentication
- Ensure proper URL encoding for special characters in tokens

raviharshicorp committed:
Fix PostgreSQL IAM auth: use pgmultiauth library approach

Based on Atlas app implementation and TFE auth.go analysis:
- Set DATABASE_URL to base connection string WITHOUT password for IAM auth
- Remove runtime token generation (pgmultiauth library handles this internally)
- Let TFE's pgmultiauth library generate IAM tokens automatically
- DATABASE_URL format: postgresql://user@host:5432/db?sslmode=require
- DATABASE_AUTH_USE_AWS_IAM=true triggers IAM authentication in pgmultiauth

raviharshicorp committed:
Fix DATABASE_URL construction: remove duplicate port specification

database_host already includes ':5432' port from RDS endpoint, so don't
add ':5432' again. This was causing malformed DATABASE_URL which resulted
in 'DATABASE_URL: null' in compose.yaml.

Correct format: postgresql://user@host:5432/db?params

raviharshicorp committed:
Remove TFE_DATABASE_PASSWORD entirely for IAM authentication

- Set TFE_DATABASE_PASSWORD to null when IAM auth is enabled
- Add TFE_DATABASE_USE_INSTANCE_PROFILE for AWS instance profile
- Filter out null values so TFE_DATABASE_PASSWORD doesn't appear in compose at all
- This prevents TFE config validation from checking for password when using IAM

raviharshicorp committed:
Fix PostgreSQL IAM auth by adding TFE_DATABASE_PASSWORDLESS environment variables

- Add TFE_DATABASE_PASSWORDLESS_AWS_USE_INSTANCE_PROFILE mapping to fix TFE config validation
- Add TFE_DATABASE_PASSWORDLESS_AWS_REGION for completeness
- These environment variables map to Database.Passwordless struct fields in TFE config
- Resolves 502 error by ensuring TFE validation recognizes IAM auth is enabled

raviharshicorp committed:
Fix PostgreSQL passwordless authentication environment variables

- Set TFE_DATABASE_PASSWORD to empty string for IAM auth (required by TFE)
- Add DATABASE_AUTH_AWS_SERVICE_NAME=rds-db for Atlas IAM provider
- Fix DATABASE_URL format with empty password placeholder (:@)
- Remove duplicate port from DATABASE_URL (host already includes port)

These fixes address the 502 errors by ensuring Atlas can properly:
1. Initialize AWS IAM token provider with correct service name
2. Parse database connection parameters from properly formatted URL
3. Generate IAM tokens for PostgreSQL RDS connections

raviharshicorp committed:
Remove unwanted Atlas environment variables for PostgreSQL IAM auth

Remove DATABASE_AUTH_* and DATABASE_URL variables as they should not be
required for PostgreSQL IAM authentication. The TFE core system should
handle IAM authentication using only the TFE_DATABASE_* variables.

raviharshicorp committed:
Add automated PostgreSQL IAM user setup during instance startup

- Add SSM document execution in user_data template
- Execute SSM command during TFE FDO startup to create IAM user automatically
- Add postgres_iam_setup_ssm_document variable to tfe_init module
- Wait for command completion and log results
- Prevents 502 errors caused by missing IAM user in PostgreSQL database

raviharshicorp committed:
Optimize user_data script to fit AWS 16KB limit

- Compress PostgreSQL IAM setup script significantly
- Reduce verbose logging to minimize user_data size
- Keep essential functionality: SSM execution and status monitoring
- Use shorter variable names and compact loops

raviharshicorp committed:
Fix cloud-init script failures for PostgreSQL IAM setup

- Add IMDSv2 token support with IMDSv1 fallback for metadata access
- Fix bash syntax errors in conditional statements and loops
- Add proper error handling and validation for SSM commands
- Use while loops instead of bash ranges for compatibility
- Add metadata debugging output for troubleshooting

raviharshicorp committed:
Replace SSM approach with direct PostgreSQL IAM user creation

- Remove complex SSM command execution that was causing failures
- Create IAM user directly in user_data script using psql
- Install PostgreSQL client and handle database connection inline
- Use template variables for database connection parameters
- Eliminate dependency on SSM document and metadata service
- Simpler, more reliable approach for IAM user creation

raviharshicorp committed:
Ultra-compact PostgreSQL IAM user setup to fit 16KB limit

- Compress entire IAM setup into 4 concise lines
- Remove all verbose logging and error handling to save space
- Use compact for loop with seq instead of while loops
- Combine all SQL grants into single psql command
- Essential functionality preserved in minimal footprint

raviharshicorp committed:
Fix PostgreSQL IAM user creation with better logging and error handling

- Add detailed logging for each step of the PostgreSQL setup
- Add retry logic for PostgreSQL client installation
- Add proper connectivity testing with version check
- Add user verification after creation
- Add explicit error handling with exit 1 on failure
- Add more descriptive log messages to debug cloud-init issues

raviharshicorp committed:
Remove PostgreSQL IAM user setup from user_data script

PostgreSQL IAM user creation is now handled by null_resource
in terraform-aws-terraform-enterprise database module instead
of user_data script. This provides better error handling and
ensures the user is created before TFE container starts.

raviharshicorp committed:
Restore PostgreSQL IAM user creation in user_data script

- Add robust PostgreSQL IAM user creation back to EC2 user_data
- Include proper error handling and logging
- EC2 instances have network access to RDS unlike CI machines
- Compact but complete implementation for AWS size limits

raviharshicorp committed:
Create ultra-compact PostgreSQL IAM user setup for AWS 16KB limit

- Reduced from ~50 lines to ~15 lines of PostgreSQL setup
- Install specific postgresql-client-16 package
- Compact single-line SQL command for IAM user creation
- Reduced logging but kept essential status messages
- File now 10KB vs previous ~12KB to ensure AWS compliance
- Maintains all required functionality for IAM user creation

raviharshicorp committed:
Remove postgres_iam_setup_ssm_document variable and references

- Remove unused variable from tfe_init module
- Update user_data template condition to use database_iam_username
- Clean up SSM document approach in favor of direct automation

raviharshicorp committed:
Clean up postgres passwordless implementation

- Fix terraform formatting issues
- Code follows best practices and naming conventions

raviharshicorp committed:
Fix tflint warnings: remove 8 unused variable declarations

- Remove unused Redis passwordless AWS variables:
  - redis_passwordless_aws_use_iam
  - redis_passwordless_aws_region
  - redis_passwordless_aws_host_name
- Remove unused Redis Sidekiq variables:
  - redis_sidekiq_host
  - redis_sidekiq_user
  - redis_sidekiq_password
  - redis_sidekiq_use_tls
  - redis_sidekiq_use_auth

All variables were declared but never referenced in the codebase.
Fixes terraform_unused_declarations warnings from tflint.

raviharshicorp committed:
Fix tflint warnings: remove 2 unused database IAM variable declarations

raviharshicorp committed:
Fix invalid TFE database environment variable

- Remove invalid TFE_DATABASE_USE_INSTANCE_PROFILE variable
- Keep only the correct TFE_DATABASE_PASSWORDLESS_AWS_USE_INSTANCE_PROFILE
- TFE_DATABASE_USE_INSTANCE_PROFILE is not a valid TFE environment variable
- Eliminates duplicate and invalid configuration

Line 19: TFE_DATABASE_USE_INSTANCE_PROFILE (invalid) → removed
Line 21: TFE_DATABASE_PASSWORDLESS_AWS_USE_INSTANCE_PROFILE (valid) → kept

raviharshicorp committed:
Restore Explorer database passwordless Azure variables

- Add back explorer_database_passwordless_azure_use_msi variable
- Add back explorer_database_passwordless_azure_client_id variable
- These were incorrectly removed as they are for Explorer database, not primary database
- Our scope is limited to primary database changes only

Reverts scope creep - Explorer database variables should remain untouched.

kkavish commented on modules/runtime_container_engine_config/database_config.tf:
```diff
@@ -3,8 +3,9 @@
   database = {
-    TFE_DATABASE_USER                         = var.database_user
-    TFE_DATABASE_PASSWORD                     = var.database_password
+    TFE_DATABASE_USER = var.database_user
```
please run `terraform fmt` before pushing commits.

kkavish commented on modules/runtime_container_engine_config/database_config.tf:
```diff
@@ -14,16 +15,18 @@ locals {
     TFE_DATABASE_PASSWORDLESS_AZURE_CLIENT_ID = var.database_passwordless_azure_client_id
+    # AWS IAM authentication for PostgreSQL passwordless
+    TFE_DATABASE_PASSWORDLESS_AWS_USE_INSTANCE_PROFILE = var.database_passwordless_aws_use_iam
+    TFE_DATABASE_PASSWORDLESS_AWS_REGION               = var.database_passwordless_aws_region
```
Maybe we can do it in a separate PR, but I am thinking the variables `database_passwordless_aws_region` and `s3_region` have the same values. Does it makes sense to merge them into one?

kkavish commented on modules/tfe_init/templates/aws.ubuntu.docker.tfe.sh.tpl:
```diff
@@ -183,4 +183,21 @@ mkdir -p $tfe_dir
 
+%{ if database_iam_username != null && database_iam_username != "" ~}
+echo "[$(date +"%FT%T")] Setting up PostgreSQL IAM user" | tee -a $log_pathname
+sudo apt-get update -qq && sudo apt-get install -y postgresql-client-16 >/dev/null 2>&1
```
Looks a little fragile, what happens when we use pg 15 or 17 as in compatibility tests?

ajmera-naman commented on modules/runtime_container_engine_config/database_config.tf:
```diff
@@ -3,8 +3,9 @@
-    TFE_DATABASE_PASSWORD                     = var.database_password
+    TFE_DATABASE_USER = var.database_user
+    # For IAM authentication, set empty password but ensure the variable exists
+    TFE_DATABASE_PASSWORD                     = var.database_passwordless_aws_use_iam ? "" : var.database_password
```
We can directly send "" for var.database_password

ajmera-naman commented on modules/runtime_container_engine_config/database_config.tf:
```diff
@@ -14,16 +15,18 @@ locals {
   }
-  database_configuration = local.disk ? {} : local.database
+  # Filter out null values so they don't appear in the compose file at all
+  database_configuration = local.disk ? {} : { for k, v in local.database : k => v if v != null }
```
Is this needed

ajmera-naman commented on modules/runtime_container_engine_config/database_config.tf:
```diff
@@ -14,16 +15,18 @@ locals {
-    TFE_EXPLORER_DATABASE_USER                         = var.explorer_database_user
-    TFE_EXPLORER_DATABASE_PASSWORD                     = var.explorer_database_password
-    TFE_EXPLORER_DATABASE_PARAMETERS                   = var.explorer_database_parameters
-    TFE_EXPLORER_DATABASE_PASSWORDLESS_AZURE_USE_MSI   = var.explorer_database_passwordless_azure_use_msi
```
Why are we removing this

ajmera-naman commented on modules/runtime_container_engine_config/database_config.tf:
```diff
@@ -14,16 +15,18 @@ locals {
-    TFE_EXPLORER_DATABASE_PASSWORD                     = var.explorer_database_password
-    TFE_EXPLORER_DATABASE_PARAMETERS                   = var.explorer_database_parameters
-    TFE_EXPLORER_DATABASE_PASSWORDLESS_AZURE_USE_MSI   = var.explorer_database_passwordless_azure_use_msi
-    TFE_EXPLORER_DATABASE_PASSWORDLESS_AZURE_CLIENT_ID = var.explorer_database_passwordless_azure_client_id
```
same as above

kkavish commented on modules/tfe_init/templates/aws.ubuntu.docker.tfe.sh.tpl:
```diff
@@ -183,4 +183,21 @@ mkdir -p $tfe_dir
+for i in $(seq 1 20); do
+if psql -h "${database_host}" -U "${admin_database_username}" -d "${database_name}" -c "SELECT 1;" >/dev/null 2>&1; then
+echo "DB connected on attempt $i" | tee -a $log_pathname
+psql -h "${database_host}" -U "${admin_database_username}" -d "${database_name}" -c "DO \$\$ BEGIN IF NOT EXISTS (SELECT 1 FROM pg_roles WHERE rolname = '${database_iam_username}') THEN CREATE USER \"${database_iam_username}\" WITH LOGIN; GRANT rds_iam TO \"${database_iam_username}\"; GRANT CONNECT ON DATABASE \"${database_name}\" TO \"${database_iam_username}\"; GRANT USAGE, CREATE ON SCHEMA pub ...
+echo "IAM user ${database_iam_username} ready" | tee -a $log_pathname
```
Is this what we suggest to our clients as well? To give admin privileges to the IAM user on all schemas? The reason I ask - release tests should be similar in deployment to what we suggest our clients, if the privileges we suggest change some time in the future the test should break.

Also, have we tried doing this via IAM roles and policies? If not, could we first try setting up that manually and if it works then incorporate it here?

kkavish commented on modules/tfe_init/templates/tfe.sh.tpl:
```diff
@@ -187,6 +187,95 @@ mkdir -p $tfe_dir
+    
+    # Create IAM user
+    echo "[$(date +"%FT%T")] [Terraform Enterprise] Creating IAM user in PostgreSQL..." | tee -a $log_pathname
+    PGPASSWORD="$admin_password" psql -h "$db_endpoint" -U "$admin_user" -d "$db_name" -v ON_ERROR_STOP=1 << EOF
+DO \$\$
+BEGIN
+    -- Check if user exists
+    IF NOT EXISTS (SELECT 1 FROM pg_roles WHERE rolname = '$iam_user') THEN
+        -- Create the IAM user
+        CREATE USER "$iam_user";
+        -- Grant rds_iam role (this role exists automatically in RDS PostgreSQL with IAM auth enabled)
+        GRANT rds_iam TO "$iam_user";
+        -- Grant necessary database permissions
+        GRANT CONNECT ON DATABASE "$db_name" TO "$iam_user";
+        GRANT USAGE ON SCHEMA public TO "$iam_user";
+        GRANT CREATE ON SCHEMA public TO "$iam_user";
+        -- Grant table permissions
+        GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO "$iam_user";
+        GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO "$iam_user";
+        -- Grant default privileges for future objects
+        ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT ALL ON TABLES TO "$iam_user";
+        ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT ALL ON SEQUENCES TO "$iam_user";
+        RAISE NOTICE 'Successfully created IAM user: $iam_user';
+    ELSE
+        RAISE NOTICE 'IAM user already exists: $iam_user';
+    END IF;
```
What is the difference between the things we are doing here and in `modules/tfe_init/templates/aws.ubuntu.docker.tfe.sh.tpl`?

ajmera-naman commented on modules/tfe_init/main.tf:
```diff
@@ -100,7 +100,22 @@ locals {
       redis_bootstrap_ca_pathname   = local.redis_bootstrap_ca_pathname
 
+      # Database configuration for templates
+      Database = {
```
why are we adding this

ajmera-naman reviewed (changes requested):
Tests are failing

raviharshicorp committed:
Address PR feedback for PostgreSQL passwordless configuration

Fix multiple issues based on code review comments:

1. Set TFE_DATABASE_PASSWORD to empty string directly (ajmera-naman feedback)
2. Remove unnecessary null filtering logic - not needed for passwordless setup
3. Restore explorer database variables that were accidentally removed
4. Remove duplicate TFE_EXPLORER_DATABASE_PARAMETERS line
5. Run terraform fmt to fix formatting issues (kkavish feedback)
6. Simplify database_configuration assignment

This addresses all PR review comments while focusing the module on PostgreSQL
IAM passwordless authentication and maintaining explorer database functionality.

raviharshicorp commented:
Closing this as this repo is archived. New PR in FDO repo: https://github.com/hashicorp/terraform-terraform-enterprise-fdo/pull/68
